src/proc/src/linux.rs

```diff
@@ -89,127 +92,135 @@ extern "C" fn iterate_cb(info: *mut dl_phdr_info, _size: size_t, data: *mut c_vo
-                            && usize::cast_from(nh.n_namesz) == GNU_NOTE_NAME.len()
-                        {
-                            // Justification: since `n_namesz` is 4, the name is a four-byte value.
-                            #[allow(clippy::as_conversions)]
-                            let p_name =
-                                (offset + std::mem::size_of::<NoteHeader>()) as *const [u8; 4];
-
-                            // SAFETY: since `n_namesz` is 4, the name is a four-byte value.
-                            assert_pointer_valid(p_name);
-                            let name = unsafe { p_name.as_ref() }.expect("pointer is valid");
-
-                            if name == GNU_NOTE_NAME {
-                                // We found what we're looking for!
-                                // Justification: simple pointer arithmetic
-                                #[allow(clippy::as_conversions)]
-                                let p_desc = (p_name as usize + 4) as *const u8;
-
-                                // SAFETY: This is the documented meaning of `n_descsz`.
-                                assert_pointer_valid(p_desc);
-                                let desc = unsafe {
-                                    std::slice::from_raw_parts(
-                                        p_desc,
-                                        usize::cast_from(nh.n_descsz),
-                                    )
-                                };
-
-                                found_build_id = Some(BuildId(desc.to_vec()));
-                                break 'outer;
-                            }
-                        }
-                        offset = offset
-                            + std::mem::size_of::<NoteHeader>()
-                            + align_up::<ELF_NOTE_STRING_ALIGN>(usize::cast_from(nh.n_namesz))
-                            + align_up::<ELF_NOTE_STRING_ALIGN>(usize::cast_from(nh.n_descsz));
+                        let p_desc = (p_name as usize + 4) as *const u8;
+
+                        // SAFETY: This is the documented meaning of `n_descsz`.
+                        assert_pointer_valid(p_desc);
+                        let desc = unsafe {
+                            std::slice::from_raw_parts(p_desc, usize::cast_from(nh.n_descsz))
+                        };
+
+                        build_id = Some(BuildId(desc.to_vec()));
+                        break;
                     }
                 }
-            }
-            if let Some(build_id) = found_build_id {
-                ve.insert(build_id);
+                offset = offset
+                    + std::mem::size_of::<NoteHeader>()
+                    + align_up::<ELF_NOTE_STRING_ALIGN>(usize::cast_from(nh.n_namesz))
+                    + align_up::<ELF_NOTE_STRING_ALIGN>(usize::cast_from(nh.n_descsz));
             }
         }
     }
-    0
+
+    let objects = state.result.as_mut().expect("we return early on errors");
```

Comment:
What will happen if we panic here? Is it okay in Rust to panic in a function that's called as a callback by some C function?

Reply:
It's a good question! I searched around a bit and this seems like a good summary: https://users.rust-lang.org/t/passing-callbacks-to-c-panic/91080/20

TLDR: Since Rust 1.71 `extern "C"` functions always abort on panic. But in Mz we are configured to abort on panic anyway, so would have done that even in earlier Rust versions.